### PR TITLE
Linkmgrd subscribing State DB route event 

### PR DESCRIPTION
**Pull request by zjswhhh**
### Description of PR

Summary:
Fixes # (issue)

This PR is to make linkmgrd subscribe events from ROUTE_TABLE in State DB, and react accordingly:  
- If any of the two default route state appears to be 'na', linkmgrd should switch to standby. 
- If both are 'ok', there will be a mux state probing and what happens next depends on linkmgrd state machine and the probing response.  

### Type of change

- [ ] Bug fix
- [ ] Test case(new/improvement)
- [x] New feature

#### What is the motivation for this PR?

To make linkmgrd subscribe state DB route event, and handle the switchovers. 

### Documentation 

Related PR: https://github.com/Azure/sonic-swss/pull/2009

### CODE REVIEW

File: src/MuxManager.cpp [hunk 1]
```diff
@@ -234,6 +234,33 @@ void MuxManager::processProbeMuxState(const std::string &portName, const std::st
...
+    MUXLOGINFO(boost::format("%s: default route state: %s") % address % routeState);
+
+    if (address.is_v4()) {
+        mIpv4DefaultRouteState = routeState;
+    } else {
+        mIpv6DefaultRouteState = routeState;
+    }
+
+    std::string nextState = "na";
+    if (mIpv4DefaultRouteState == "ok" && mIpv6DefaultRouteState == "ok") {
```
Comment (yxieca) [inline src/MuxManager.cpp:253]: @lguohan do we have concern if either IPv4 or IPv6 BGP is not configured? Can we assume we will have them both eventually?

Comment (zjswhhh) [reply src/MuxManager.cpp]: Updated per discussion in Gemini sync-up meeting.

File: src/link_manager/LinkManagerStateMachine.cpp [hunk 2]
```diff
@@ -815,6 +815,29 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
...
+// 
+// ---> handleDefaultRouteStateNotification(const std::string &routeState);
+// 
+// handle default route state notification from routeorch
+//
+void LinkManagerStateMachine::handleDefaultRouteStateNotification(const std::string &routeState)
+{
+    MUXLOGWARNING(boost::format("%s: state db default route state: %s") % mMuxPortConfig.getPortName() % routeState);
+
+    if(mComponentInitState.test(MuxStateComponent)) {
```
Comment (yxieca) [inline src/link_manager/LinkManagerStateMachine.cpp:827]: add space after if

File: src/link_manager/LinkManagerStateMachine.cpp [hunk 3]
```diff
@@ -815,6 +815,29 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
...
+        if (ms(mCompositeState) == mux_state::MuxState::Label::Active &&
+            routeState == "na") {
```
Comment (yxieca) [inline src/link_manager/LinkManagerStateMachine.cpp:829]: I suggest putting '{' on a new line, especially when if condition expands multiple lines. new line helps to identify the beginning of the if body.

File: src/MuxManager.cpp [hunk 4]
```diff
@@ -234,6 +234,34 @@ void MuxManager::processProbeMuxState(const std::string &portName, const std::st
...
+    }
+
+    std::string nextState = "na";
+    // For now we only need IPv4 default route state, this logic can be updated in the future.  
+    if (mIpv4DefaultRouteState == "ok") {
+        nextState = "ok";
+    }
+
+    PortMapIterator portMapIterator = mPortMap.begin();
+    while(portMapIterator != mPortMap.end()) {
```
Comment (yxieca) [inline src/MuxManager.cpp:259]: space after while please.

File: src/link_manager/LinkManagerStateMachine.cpp (hunk 2 above)
Comment (zjswhhh) [reply src/link_manager/LinkManagerStateMachine.cpp]: Updated accordingly.

File: src/link_manager/LinkManagerStateMachine.cpp (hunk 3 above)
Comment (zjswhhh) [reply src/link_manager/LinkManagerStateMachine.cpp]: Updated accordingly.

File: src/MuxManager.cpp (hunk 4 above)
Comment (zjswhhh) [reply src/MuxManager.cpp]: Updated accordingly.

File: src/link_manager/LinkManagerStateMachine.cpp [hunk 5]
```diff
@@ -815,6 +815,28 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
...
+// ---> handleDefaultRouteStateNotification(const std::string &routeState);
+// 
+// handle default route state notification from routeorch
+//
+void LinkManagerStateMachine::handleDefaultRouteStateNotification(const std::string &routeState)
+{
+    MUXLOGWARNING(boost::format("%s: state db default route state: %s") % mMuxPortConfig.getPortName() % routeState);
+
+    if (mComponentInitState.test(MuxStateComponent)) {
+        if (ms(mCompositeState) == mux_state::MuxState::Label::Active && routeState == "na") {
```
Comment (prsunny) [inline src/link_manager/LinkManagerStateMachine.cpp:828]: I'm thinking we should check for `!= Standby`. What if mux is in a transient state and we got this event? What do you think?

File: src/link_manager/LinkManagerStateMachine.cpp [hunk 6]
```diff
@@ -815,6 +815,28 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
...
+            CompositeState nextState = mCompositeState;
+            enterLinkProberState(nextState, link_prober::LinkProberState::Wait);
+            switchMuxState(nextState, mux_state::MuxState::Label::Standby);
```
Comment (prsunny) [inline src/link_manager/LinkManagerStateMachine.cpp:831]: How does it behave if both ToRs have default route as "na" ?

Comment (zjswhhh) [reply src/link_manager/LinkManagerStateMachine.cpp:831]: Both ToRs will switch to standby, the one switches first will be "active" at the end. Do we expect both ToRs to be "standby" in this scenario?

File: src/link_manager/LinkManagerStateMachine.cpp (hunk 5 above)
Comment (zjswhhh) [reply src/link_manager/LinkManagerStateMachine.cpp:828]: Yes I think this is a good point. I will evaluate the behavior and update accordingly.

**Review by prsunny — APPROVED** [review]

**Comment by zjswhhh** [comment]
boost::asio::ip::make_address is unable to parse "::/0" and "0.0.0.0/0", and considering the key will always be these two strings, I updated the code and removed make_address(). Changes have been tested on virtual dual testbeds and worked as expected. 

@prsunny could you review again? Thanks!

**Pull request merged**